java/client/src/org/openqa/selenium/support/pagefactory/ByChained.java

```diff
@@ -48,10 +48,21 @@ public ByChained(By... bys) {
 
   @Override
   public WebElement findElement(SearchContext context) {
-    List<WebElement> elements = findElements(context);
-    if (elements.isEmpty())
-      throw new NoSuchElementException("Cannot locate an element using " + toString());
-    return elements.get(0);
+    if (bys.length == 0) {
+      throw new NoSuchElementException("No Bys were specified in this " + getClass().getSimpleName());
+    }
+
+    WebElement elem = null;
+    for (By by : bys) {
+      boolean isFirstIteration = elem == null;
+      if (isFirstIteration) {
+        elem = by.findElement(context);
+      } else {
+        elem = elem.findElement(by);
+      }
+    }
+    assert elem != null; // NoSuchElement should have already been thrown
```

Comment:
Hi Luke,

Thanks for your feedback.

Uhm actually NoSuchEl may be thrown by

by.findElement(context)
Or
elem.findElement(by)

So it will stop and throw as soon as any one of the bys retrieves no
elements. If I'm not mistaken!

Example.
ByChained(by1, by2, by3)
By1 retrieves 3 elements. I save the first
By2 retrieves nothing. NoSuchEl is thrown by elem.findElement(by).

Let me know how it sounds.

About the assert, no problem, I will change it.

Thanks

Alberto Scotto
http://alb-i986.me
[on Android]
Il 26/mar/2015 15:13 "Luke Inman-Semerau" <notifications@github.com> ha
scritto:

> In java/client/src/org/openqa/selenium/support/pagefactory/ByChained.java
> <https://github.com/SeleniumHQ/selenium/pull/372#discussion_r27221662>:
>
> > -      throw new NoSuchElementException("Cannot locate an element using " + toString());
> > -    return elements.get(0);
> > +    if (bys.length == 0) {
> > +      throw new NoSuchElementException("No Bys were specified in this " + getClass().getSimpleName());
> > +    }
> > +
> > +    WebElement elem = null;
> > +    for (By by : bys) {
> > +      boolean isFirstIteration = elem == null;
> > +      if (isFirstIteration) {
> > +        elem = by.findElement(context);
> > +      } else {
> > +        elem = elem.findElement(by);
> > +      }
> > +    }
> > +    assert elem != null; // NoSuchElement should have already been thrown
>
> Let's not use "assert" but rather throw an Exception explicitly.
>
> Overall, this logic re-write doesn't maintain the existing functionality.
> Namely if any of the by's do not find the element, it doesn't throw an
> Element Not Found exception, but rather continues executing all by's until
> at least one was found.
>
> I would change your for loop to look more like this:
>
> for (By by : bys) {
>   List<WebElements> elements = by.findElements(elem == null ? context : elem);
>   if (!elements.isEmpty()) {
>     elem = elements.get(0);
>   }
> }
>
> —
> Reply to this email directly or view it on GitHub
> <https://github.com/SeleniumHQ/selenium/pull/372/files#r27221662>.
>